security/abac/role/src/main/java/io/helidon/security/abac/role/RoleValidator.java

```diff
@@ -84,37 +88,51 @@ public RoleConfig fromConfig(Config config) {
-    public RoleConfig fromAnnotations(List<? extends Annotation> annotations) {
+    public RoleConfig fromAnnotations(EndpointConfig endpointConfig) {
         RoleConfig.Builder builder = RoleConfig.builder();
 
-        for (Annotation annotation : annotations) {
-            if (annotation instanceof RolesAllowed) {
-                // these are user roles by default
-                builder.addRoles(Arrays.asList(((RolesAllowed) annotation).value()));
-            } else if (annotation instanceof Roles) {
-                // these are configured
-                Roles r = (Roles) annotation;
-                if (r.subjectType() == SubjectType.USER) {
-                    builder.addRoles(Arrays.asList(r.value()));
-                } else {
-                    builder.addServiceRoles(Arrays.asList(r.value()));
+        for (EndpointConfig.AnnotationScope value : EndpointConfig.AnnotationScope.values()) {
+            List<Annotation> annotations = new ArrayList<>();
+            for (Class<? extends Annotation> annotation : supportedAnnotations()) {
+                annotations.addAll(endpointConfig.combineAnnotations(annotation, value));
+            }
+            for (Annotation annotation : annotations) {
+                if (annotation instanceof RolesAllowed) {
+                    // these are user roles by default
+                    builder.addRoles(Arrays.asList(((RolesAllowed) annotation).value()));
+                    builder.permitAll(false);
+                    builder.denyAll(false);
+                } else if (annotation instanceof Roles) {
+                    // these are configured
+                    Roles r = (Roles) annotation;
+                    if (r.subjectType() == SubjectType.USER) {
+                        builder.addRoles(Arrays.asList(r.value()));
+                    } else {
+                        builder.addServiceRoles(Arrays.asList(r.value()));
+                    }
+                    builder.permitAll(false);
+                    builder.denyAll(false);
+                } else if (annotation instanceof PermitAll) {
+                    builder.permitAll(true);
+                    builder.denyAll(false);
+                } else if (annotation instanceof DenyAll) {
+                    builder.permitAll(false);
+                    builder.denyAll(true);
                 }
             }
         }
-
         return builder.build();
     }
 
-    @Override
-    public RoleConfig combine(RoleConfig parent, RoleConfig child) {
-        return RoleConfig.builder()
-                .addRoles(parent.userRolesAllowed())
-                .addRoles(child.userRolesAllowed())
-                .build();
-    }
-
     @Override
     public void validate(RoleConfig config, Errors.Collector collector, ProviderRequest request) {
+        if (config.permitAll()) {
+            return;
+        }
+        if (config.denyAll()) {
+            collector.fatal(this, "This target is not accessible by any means!");
```

Comment:
This message is useless. Please use something like `Access denied by DenyAll`

Reply:
Message edited according to your suggestion.